src/ast/passes/map_sugar.cpp

```diff
@@ -0,0 +1,274 @@
+#include <unordered_map>
+
+#include "ast/ast.h"
+#include "ast/passes/map_sugar.h"
+#include "ast/visitor.h"
+
+namespace bpftrace::ast {
+
+namespace {
+
+class MapDefaultKey : public Visitor<MapDefaultKey> {
+public:
+  explicit MapDefaultKey(ASTContext &ast) : ast_(ast) {};
+
+  template <typename T>
+  T *replace(T *node, [[maybe_unused]] void *result)
+  {
+    if constexpr (std::is_same_v<T, Expression>) {
+      // Replace with an indexed map. Note that we don't visit for calls that
+      // are exempt from this. This applies to all expressions in the tree,
+      // including `lhist` and `hist`, which are treated in a special way
+      // subsequently.
+      if (auto *map = dynamic_cast<Map *>(node)) {
+        auto *index = ast_.make_node<Integer>(0, Location(map->loc));
+        return ast_.make_node<MapAccess>(map, index, Location(map->loc));
+      }
+      return node;
+    } else if constexpr (std::is_same_v<T, Statement>) {
+      // Replace with a statement that has the default index, in the same way as
+      // above. This will be type-checked during semantic analysis.
+      if (auto *map = dynamic_cast<AssignScalarMapStatement *>(node)) {
+        auto *index = ast_.make_node<Integer>(0, Location(map->loc));
+        return ast_.make_node<AssignMapStatement>(
+            map->map, index, map->expr, Location(node->loc));
+      }
+      return node;
+    } else {
+      return Visitor<MapDefaultKey>::replace<T>(node, result);
+    }
+  }
+
+  using Visitor<MapDefaultKey>::visit;
+  void visit(Call &call);
+  void visit(For &for_loop);
+  void visit(Map &map);
+  void visit(MapAccess &acc);
+  void visit(AssignScalarMapStatement &assign);
+  void visit(AssignMapStatement &assign);
+  void check(Map &map, bool indexed);
+
+  MapMetadata metadata()
```

Comment:
why move the metadata out of here? maybe the function should be named `extract_metadata`?

Reply:
I can just make it public and move it out from the pass.

Reply:
Done.